File: samples/java-protobuf-tracing/src/main/java/com/example/ControllerAction.java
package com.example;

import com.google.protobuf.Empty;
import io.opentelemetry.api.trace.Span;
import io.opentelemetry.api.trace.StatusCode;
import io.opentelemetry.api.trace.Tracer;
import kalix.javasdk.action.ActionCreationContext;

import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.util.concurrent.CompletableFuture;

public class ControllerAction extends AbstractControllerAction {

  String url = "https://jsonplaceholder.typicode.com/posts";
  HttpClient httpClient = HttpClient.newHttpClient();
  public ControllerAction(ActionCreationContext creationContext) {}

  // tag::create-close-span[]
  @Override
  public Effect<ControllerActionApi.MessageResponse> callAsyncEndpoint(Empty empty) {
    // tag::get-tracer[]
    Tracer tracer = actionContext().getTracer();
    // end::get-tracer[]

    Span span = tracer
              .spanBuilder(url+"/{}")
              .setParent(actionContext().metadata().traceContext().asOpenTelemetryContext())// <1>
              .startSpan() // <2>
              .setAttribute("post", "1");// <3>

    CompletableFuture<ControllerActionApi.MessageResponse> asyncComputation = callAsyncService()
      .whenComplete((response, ex) -> {
        if (ex != null) {
          span
              .setStatus(StatusCode.ERROR, ex.getMessage())// <4>
              .end();// <5>
        } else {
          span
              .setAttribute("result", response.body().title)// <3>
              .end();// <5>
        }
      })
      .thenApply(response ->
          ControllerActionApi.MessageResponse.newBuilder().setMessage(response.body().title).build()
      );
    return effects().asyncReply(asyncComputation);
  }
  // end::create-close-span[]


  private CompletableFuture<HttpResponse<Post>> callAsyncService(){
    HttpRequest httpRequest = HttpRequest.newBuilder()
            .uri(URI.create(url+"/1"))
            .build();
    //Async call to external service
    return httpClient.sendAsync(httpRequest,
            new JsonResponseHandler<>(Post.class));
  }

  public record Post(String userId, String id, String title, String body) {}



}

File: samples/scala-protobuf-tracing/src/main/scala/com/example/ControllerAction.scala
package com.example

import com.google.protobuf.empty.Empty
import io.opentelemetry.api.trace.{StatusCode, Tracer}
import io.opentelemetry.context.Scope
import kalix.scalasdk.action.{Action, ActionCreationContext}
import sttp.client4.quick.RichRequest
import sttp.client4.{Response, _}

import scala.concurrent.Future
import scala.util.{Failure, Success}

class ControllerAction(creationContext: ActionCreationContext) extends AbstractControllerAction {

  val url = "https://jsonplaceholder.typicode.com/posts/1"

  override def callAsyncEndpoint(empty: Empty): Action.Effect[MessageResponse] = {
    val tracerOpt = actionContext.getOpenTelemetryTracer
    tracerOpt match {
      case Some(tracer) =>
        val responseBody = callAsync(tracer)
        effects.asyncReply(responseBody)
      case None =>
        val responseBody = callAsync()
        effects.asyncReply(responseBody)
    }
  }

  private def callAsync(tracer: Tracer): Future[MessageResponse] = {
    val span = tracer
      .spanBuilder("loreipsumendpoint")
      .setParent(actionContext.metadata.traceContext.asOpenTelemetryContext)
      .startSpan()
    val scope: Scope = span.makeCurrent()
    try {
      val responseBody: Future[MessageResponse] = callAsync()
      responseBody.onComplete {
        case Success(response) => span.setAttribute("result", response.message)
        case Failure(exception) => span.setStatus(StatusCode.ERROR, exception.getMessage)
      }
      responseBody
    } finally {
      span.end()
      scope.close()
    }
  }

  private def callAsync(): Future[MessageResponse] = {
    val request = basicRequest.get(uri"$url")
    val response: Future[Response[Either[String, String]]] =
      request.send(Main.backend)
    response.map { response =>
      response.body match {
        case Left(resEx) =>
          MessageResponse(resEx)
        case Right(value) =>
          MessageResponse(value)
      }
    }
  }

}
